cudax/include/cuda/experimental/__memory_resource/shared_resource.cuh

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__MEMORY_RESOURCE_SHARED_RESOURCE_H
+#define _CUDAX__MEMORY_RESOURCE_SHARED_RESOURCE_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+// If the memory resource header was included without the experimental flag,
+// tell the user to define the experimental flag.
+#if defined(_CUDA_MEMORY_RESOURCE) && !defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+#  error "To use the experimental memory resource, define LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE"
+#endif
+
+// cuda::mr is unavable on MSVC 2017
+#if defined(_CCCL_COMPILER_MSVC_2017)
+#  error "The shared_resource header is not supported on MSVC 2017"
+#endif
+
+#if !defined(LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE)
+#  define LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE
+#endif
+
+#include <cuda/__memory_resource/resource.h>
+#include <cuda/std/__type_traits/is_swappable.h>
+#include <cuda/std/__type_traits/type_set.h>
+#include <cuda/std/__utility/move.h>
+
+#include <cuda/experimental/__memory_resource/any_resource.cuh>
+
+#include <memory>
+
+namespace cuda::experimental::mr
+{
+
+//! @rst
+//! .. _cudax-memory-resource-shared-resource:
+//!
+//! Resource wrapper to share ownership of a resource
+//! --------------------------------------------------
+//!
+//! ``shared_resource`` holds a reference counted :ref:`any_resource <cudax-memory-resource-basic-any-resource>`.
+//! This allows the user to pass a resource around with reference semantics while avoiding lifetime issues.
+//!
+//! @endrst
+template <class... _Properties>
+class shared_resource
+{
+private:
+  ::std::shared_ptr<::cuda::experimental::mr::any_resource<_Properties...>> __resource_;
+
+  template <class... _OtherProperties>
+  static constexpr bool __properties_match =
+    _CUDA_VSTD::__type_set_contains<_CUDA_VSTD::__make_type_set<_OtherProperties...>, _Properties...>;
+
+  template <class _Resource>
+  static constexpr bool __different_resource =
+    !_CCCL_TRAIT(::cuda::std::is_same, ::cuda::std::remove_cvref_t<_Resource>, shared_resource);
+
+public:
+  //! @brief Constructs a \c shared_resource from a type that satisfies the \c resource concept as well as all
+  //! properties.
+  //! @param __res The resource to be wrapped within the \c shared_resource.
+  _LIBCUDACXX_TEMPLATE(class _Resource)
+  _LIBCUDACXX_REQUIRES(
+    __different_resource<_Resource> _LIBCUDACXX_AND(_CUDA_VMR::resource_with<_Resource, _Properties...>))
+  shared_resource(_Resource&& __res)
```

Comment:
This only supports movable resources.  I think we want to be able to support (existing) non-movable resource types. If you provided a `make_shared_resource<Resource>(args...)` that might be sufficient.

Reply:
Currently there is no way to do this, because neither `resource_ref` nor `any_resource` support that usecase.

Also this should work fine with prvalues due to copy elision

Reply:
> neither `resource_ref` nor `any_resource` support that usecase

right, that's what i was saying above. `any_resource` is the wrong thing here. we need `make_shared_resource<Resource>(args...)` that dynamically allocates the resource and ref-counts it. the type in this PR does not meet RMM's needs.